[Ho Chi Thanh] iP

ho-thanh-und opened the pull request:
# Him
> "Your mind is for having ideas, not holding them." – David Allen

Him frees your mind of having to remember things you need to do. It's,

- **text-based**
- **easy to learn**
- **QUITE fast to use (maybe)**

## Getting Started

All you need to do is,

1. download it from [here](https://github.com/[ho-thanh-und/ip](https://github.com/ho-thanh-und/ip))
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

## Features:
- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders *(coming to you real soon)*

## For those who enjoy coding in Java (why?) 
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        new Him("./data/Tasks.txt").run();
    }
}
```
## Remarks
**Would love to receive suggestions to improve my project! Good luck for the rest of CS2103T!**

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ho-thanh-und committed:
For Level-0:
- Changed the Duke.java file to Him.java (will change itf)
- Added a Ui class which consists of two methods
- Used the Ui class in the Him.java file to greet and farewell users

ho-thanh-und committed:
- Removed Ui.java due to oversight
- Reverted to simple print statements and will work on from there

ho-thanh-und committed:
- Removed Ui.java due to oversight
- Reverted to simple print statements and will work on from there

ho-thanh-und committed:
Merge remote-tracking branch 'origin/master'

ho-thanh-und committed:
- Imported Scanner to retrieve input/output
- Allows echoing: For whatever input the program receives, it will regurgitate it. If it's a "bye", the program will exit

Mistakes occured:
- The input variable of type String should be nested inside the while loop as to avoid an infinite loop
- Similarly, it was necessary to add a while loop as to make the functionality of the if block work, or else it would also encounter an infinite loop

ho-thanh-und committed:
- Added listing functionality to list out added tasks from the "todos" list

Mistakes occurred
- Not really a mistake but System.out.printf seems less wordy compared to Systems.out.print(Stringformat...))
- Realized had not removed Ui.java due to some bugs occurred previously

ho-thanh-und committed:
- Added Task.java
  + Represents tasks
  + Has two methods to mark tasks as done or undone
  + And a toString method to format output

- Added more functionality for the chatbot by allowing it to mark and unmark tasks which are formatted

Mistakes occurred
- Still working on how to better comment
- The previous "todos" accepted a String type, but coming into this variation, the point of marking and unmarking tasks meant initially marking tasks in "todos" from the methods in Task.java was not possible. Changing the type to Task helped fix the bug

ho-thanh-und committed:
- Added 3 child classes of Task.java to enhance variety of tasks to todos, deadlines and events

Mistakes occurred:
- Lots of repetition with these if-else blocks. Would it be better to use a switch-case instead?

ho-thanh-und committed:
- Added testing

Mistakes occurred:
- Not being able to run the .bat file. Turns out, we just need a ./runtest.bat
- Changing chatbot name had issues with running the script

ho-thanh-und committed:
- Added Exceptions to catch errors for each specific function
- A total of 4 errors were handled, including empty descriptions and invalid arguments
- Added HimException.java to handle the errors

Mistakes occurred:
- Deciding the exceptions seemed to take up some time
- The HimException.java class is never used. I think this is due to the fact that I did not separate the specific errors for it to handle

ho-thanh-und committed:
- Added the delete functionality with 2 error-handling exceptions
- Changed the output statements of the chatbot

Mistakes occurred:
- Logic behind deleting

ho-thanh-und committed:
Add Him.txt file

ho-thanh-und committed:
Add three Exception files to handle the case where the data file doesn't exist at the start

ho-thanh-und committed:
Update the Tasks subclasses

ho-thanh-und committed:
Remodify Him.java into composite methods to enhance readability and scalability

ho-thanh-und committed:
Merge pull request #1 from ho-thanh-und/branch-Level-7

Merge branch 'branch-Level-7' to the Master branch

ho-thanh-und committed:
Add Date and Time formatting for both the Deadline.java file and the Event.java file to specify time

ho-thanh-und committed:
Merge pull request #2 from ho-thanh-und/branch-Level-8

Merge branch 'branch-Level-8' to the Master Branch

ho-thanh-und committed:
Add Ui class to handle interaction with the user

ho-thanh-und committed:
Add Storage class to handle loading and saving tasks in the file

ho-thanh-und committed:
Add Storage class to handle loading and saving tasks in the file

ho-thanh-und committed:
Add Parser class to handle user commands

ho-thanh-und committed:
Update Task.java with a new constructor to facilitate defining (in)complete tasks

ho-thanh-und committed:
Update Deadline.java with a new constructor to facilitate defining (in)complete tasks (similar to the Task.java commit)

ho-thanh-und committed:
Update Event.java with a new constructor to facilitate defining (in)complete tasks (similar to the Task.java commit) and improved the toString method

ho-thanh-und committed:
Improved the toString method in Deadline.java

ho-thanh-und committed:
Update ToDo.java by adding a new constructor

ho-thanh-und committed:
Added TaskList.java

ho-thanh-und committed:
Renamed Him.txt to him.txt and improved the getPreviousTasks method in Storage.java

ho-thanh-und committed:
Updated Him.java to better follow OOP Practices

ho-thanh-und committed:
Merge pull request #3 from ho-thanh-und/branch-A-MoreOOP

Merge branch 'branch-A-MoreOOP' to the Master Branch

ho-thanh-und committed:
Divide files by packages

ho-thanh-und committed:
Merge pull request #4 from ho-thanh-und/branch-A-Packages

Merge branch 'branch-A-Packages' to the Master Branch

ho-thanh-und committed:
Merge pull request #5 from ho-thanh-und/add-gradle-support

Add gradle support

ho-thanh-und committed:
Divide files by packages

ho-thanh-und committed:
Merge pull request #6 from ho-thanh-und/branch-A-Packages

Merge branch 'branch-A-Packages' into the Master branch

ho-thanh-und committed:
Divide files by packages

ho-thanh-und committed:
Add gradle support

ho-thanh-und committed:
Merge pull request #7 from ho-thanh-und/add-gradle-support

Add gradle support

ho-thanh-und committed:
Add ToDo

ho-thanh-und committed:
Merge pull request #8 from ho-thanh-und/branch-A-Packages

Branch a packages

ho-thanh-und committed:
Fix NullPointerException error for both Deadline.java and Event.java

ho-thanh-und committed:
Add both DeadlineTest.java and ToDoTest.java to test the respective files in main

ho-thanh-und committed:
Merge pull request #9 from ho-thanh-und/branch-A-JUnit

Merge branch 'branch-A-JUnit' to the Master branch

ho-thanh-und committed:
Update build.gradle

ho-thanh-und committed:
Merge pull request #10 from ho-thanh-und/branch-A-JUnit

Update build.gradle

ho-thanh-und committed:
Add checkstyle

ho-thanh-und committed:
Remove Him.java duplicate

ho-thanh-und committed:
Improve TaskList.java to fix error on returning stream output

ho-thanh-und committed:
Modify Him.java to provide farewell message when user exits the program overall

ho-thanh-und committed:
Modified build.gradle to work with jdk 17 and created jar file

ho-thanh-und committed:
Merge pull request #11 from ho-thanh-und/branch-A-Jar

Merge branch 'branch-A-Jar' to the Master branch

ho-thanh-und committed:
Add JavaDocs to methods/constructors

ho-thanh-und committed:
Ensure Coding Standards are followed

ho-thanh-und committed:
Merge pull request #12 from ho-thanh-und/branch-A-JavaDoc

Merge branch 'branch-A-JavaDoc' to the Master branch

ho-thanh-und committed:
Merge pull request #13 from ho-thanh-und/branch-A-CodingStandard

Merge branch 'branch-A-CodingStandard' to the Master branch

ho-thanh-und committed:
Add Finding functionality

ho-thanh-und committed:
Merge pull request #14 from ho-thanh-und/branch-Level-9

Merge branch 'branch-Level-9' to the Master branch

CloudKai commented on src/main/java/him/storage/Storage.java:
```diff
@@ -0,0 +1,90 @@
+                case "E":
+                    task = new Event(data[2], data[3], getBooleanFromString(data[1]));
+                    tasks.add(task);
+                    break;
```
Should this follow the coding standard for the switch case?

CloudKai replied:
I like how the save file and print are different formats and this managed to read the different syntax into Task

CloudKai commented on src/main/java/him/ui/Ui.java:
```diff
@@ -0,0 +1,68 @@
+        print("Glad to be of service");
+    }
+
+    /**
+     * Displays an error message when the event description or time is missing.
+     */
+    public void eventErrorMsg() {
+        print("The event description or time can't be empty bro, you gotta " +
+                "use this format: event <task> at <time>");
+    }
+
+    /**
+     * Displays an error message when the deadline description or due date is missing.
+     */
+
+    public void deadlineErrorMsg() {
+        print("The deadline description or due date can't be empty bro, you gotta " +
+                "use this format: deadline <task> by <date>");
+    }
```
I like the unique Error messages you have rather than having all under a same error message

CloudKai commented on src/main/java/him/task/Deadline.java:
```diff
@@ -0,0 +1,77 @@
+                .map(localDate -> "[D]" + super.toString() +
+                        String.format(" (by: %s)", localDate.format(DateTimeFormatter.ofPattern("MMM dd yyyy"))))
+                .orElse("[D]" + super.toString() + String.format(" (by: %s)", by));
+    }
```
You could set a DateTime Parser for both Deadline and Event class

CloudKai commented on src/main/java/him/tasklist/TaskList.java:
```diff
@@ -0,0 +1,159 @@
+                .filter(task -> {
+                    return task.getDescription().toLowerCase().contains(lowerKeyword);
+                })
+                .toList();
```
I like the use of stream here able to find the matchingTasks easily

CloudKai reviewed (commented):
The codebase is very neat! I learn a lot from how you used to match with regex and using functional programming. It was a pleasure to review your PR

mervyn-teo commented on src/main/java/him/storage/Storage.java:
```diff
@@ -0,0 +1,90 @@
+     * @throws FileNotFoundException If the file does not exist.
+     */
+    public static ArrayList<Task> getPreviousTasks(String filePath) throws FileNotFoundException {
+        File file = new File("data/data.txt");
```
Method name is in camelCase, very good!

mervyn-teo commented on src/main/java/him/storage/Storage.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Deals with loading and saving tasks
+ */
+public class Storage {
```
Class names are in Pascal Casing, very nice!

mervyn-teo commented on src/main/java/him/himhome/Him.java:
```diff
@@ -0,0 +1,102 @@
+            while (true) {
+                String input = ui.getInput();
+                if (input.equalsIgnoreCase("bye")) {
+                    ui.farewellMsg();
```
Would it be better if `switch` were used here?

mervyn-teo reviewed (commented):
Nicely done! The naming of classes and variables complied with the coding standard, comments/Javadoc are direct and easy to understand!

ho-thanh-und committed:
Update build.gradle

ho-thanh-und committed:
Add fxml features

ho-thanh-und committed:
Add fxml features

ho-thanh-und committed:
Update TaskList and Ui

ho-thanh-und committed:
Add Main.java

ho-thanh-und committed:
Update Him.java

ho-thanh-und committed:
Merge pull request #15 from ho-thanh-und/branch-Level-10

Merge branch 'branch-Level-10' to the Master Branch

ho-thanh-und committed:
Update Him.java

ho-thanh-und committed:
Fix duke image not appearing when loading app by updating the setDuke() method

ho-thanh-und committed:
Fix welcome message not appearing by updating the welcome and farewell message

ho-thanh-und committed:
Fix logic error of todo display

ho-thanh-und committed:
Fix logic error of all task display. Re-add undone logic

ho-thanh-und committed:
Merge pull request #17 from ho-thanh-und/branch-Level-10

Update branch level 10

ho-thanh-und committed:
Improve code quality issues

ho-thanh-und committed:
Remove exception classes for empty tasks

Previously, we used dedicated exception classes to handle errors from tasks that are empty. Now, however, the exception handling logic has been moved into Him#getResponse(). Consequently, these exception classes
are unused and only add clutter to the codebase.

Let's remove these classes to clean up the file/folder structure.

ho-thanh-und committed:
Merge pull request #19 from ho-thanh-und/branch-A-AiAssisted

Remove exception classes for empty tasks

ho-thanh-und committed:
Merge pull request #18 from ho-thanh-und/branch-Level-10

branch-Level-10

ho-thanh-und committed:
Add Java asserts to document assumptions

We rely on certain invariants in constructors (e.g., non-empty
descriptions) and in methods (e.g., non-empty task list). Previously, these assumptions were only implied or handled less explicitly.

Using Java's built-in assert statements clarifies these assumptions. If
a precondition is violated while running with assertions enabled (-ea),
the program throws an AssertionError early. This helps in catching bugs
during development.

Let's add assert statements to:
* Task, Deadline, and Event constructors to ensure description is not empty
* fillFileWithTasks to ensure the task list is not empty

ho-thanh-und committed:
Add Java asserts to document assumptions

We rely on certain invariants in constructors (e.g., non-empty
descriptions) and in methods (e.g., non-empty task list). Previously, these assumptions were only implied or handled less explicitly.

Using Java's built-in assert statements clarifies these assumptions. If
a precondition is violated while running with assertions enabled (-ea),
the program throws an AssertionError early. This helps in catching bugs
during development.

Let's add assert statements to:
* Task, Deadline, and Event constructors to ensure description is not empty
* fillFileWithTasks to ensure the task list is not empty

ho-thanh-und committed:
Merge pull request #20 from ho-thanh-und/branch-A-Assertions

Merge branch 'branch-A-Assertions' to the Master branch

ho-thanh-und committed:
Merge pull request #21 from ho-thanh-und/master

Rectify merge of updated master branch to branch-A-CodeQuality

ho-thanh-und committed:
Detail use of streams in toString() methods in Deadline.java and Event.java for optional dates.

 This commit highlights how this program leverages Java Streams and Optional methods (filter/map) to streamline that logic.

Using Streams clarifies how we:
* Check if the date/time is valid
* Format and display the date/time if present
* Fall back to a string-based format otherwise

This approach improves readability and reduces boilerplate code when
constructing the final string representation of these tasks.

ho-thanh-und committed:
Merge pull request #22 from ho-thanh-und/branch-A-Streams

Detail use of streams in toString() methods in Deadline.java and Even…

ho-thanh-und committed:
Introduce boolean variables to avoid having conditional expressions within if statements.

The codebase previously contained inline conditional expressions within if statements, making readability difficult.

Using boolean variables help:
- Code become self-documenting
- Improves readability

ho-thanh-und committed:
Refactor command detection to use an enum-based Parser

Previously, Him relied on a long chain of if/else statements to check user commands (e.g., "done", "deadline", "event"). This approach mixed command detection with command handling, making readability difficult.

The code had duplicated logic for each new command (task).

Improve by:
- Using enums to categorize commands (e.g., BYE, LIST)
- Add parseCommand method to convert user input into an enum
- Replace the if/else chains with the enums

This abides to 'single-level-of-abstraction' principles, making sure code is readable.

ho-thanh-und committed:
Merge pull request #23 from ho-thanh-und/branch-A-CodeQuality

Merge branch 'branch-A-CodeQuality' to the Master Branch

ho-thanh-und committed:
Add "DoWithin" tasks to TaskList, Storage, and Parser

Currently, the codebase can handle todo, deadline, and event tasks,
but lacks the ability to create tasks that must be done within a
specific period.

Users often need to track tasks that have a strictly bounded start
and end date.

Let's introduce a DoWithin class and integrate it into TaskList,
Storage, and Parser so users can add tasks that must be completed
within a given period.

We follow the existing Duke-like structure to keep the logic
consistent. Storage detects lines prefixed with 'W', TaskList
handles "dowithin" commands, and Parser recognizes them as part
of ADD_TASK. This preserves the same command flow while expanding
the task types available.

ho-thanh-und committed:
Incorporate "DoWithin" tasks in Him for user interaction

Previously, Him handled "todo", "deadline", and "event" commands
but did not account for "dowithin" tasks that must be completed
within a specified period.

Users need a straightforward way to input "dowithin" commands and
receive the correct responses from the chatbot, just like other
task types.

Let's add support for "dowithin" in Him by treating it as part of
the ADD_TASK command flow. When the user types a "dowithin" command,
Him delegates to TaskList to create DoWithin tasks.

This keeps our approach consistent with the existing code structure
and ensures a seamless experience for users adding time-bounded tasks.

ho-thanh-und committed:
Merge pull request #24 from ho-thanh-und/branch-BCD-Extension

Merge branch 'branch-BCD-Extension' to the Master branch

ho-thanh-und committed:
UI improvements: add Lakers styling, set window icon, handle empty tasks

The UI uses default styling and application window icon is not customized.
The task list throws an unhandled exception when it is empty.

The interface and functionality need to change to improve visual appeal, ensure consistent branding, and prevent application crashes when the task list is empty.

Let's apply Lakers colors (purple and gold) in main-window.css and dialog-box.css.
Let's set a custom icon in the application window using stage.getIcons().add(new Image(...)).
Let's modify displayToDo to handle when the list is empty.
Let's add a Ui display to showcase latest display.

Using a Lakers-themed design aligns the UI with the brand identity.
Explicit error handling for empty tasks enhances robustness.
Customizing the window icon and tweaking layout details ensures a polished, user-friendly experience.

ho-thanh-und committed:
Merge pull request #25 from ho-thanh-und/branch-UI

UI improvements: add Lakers styling, set window icon, handle empty tasks

ho-thanh-und committed:
UI improvements: add Lakers styling, set window icon, handle empty tasks

The UI uses default styling and application window icon is not customized.
The task list throws an unhandled exception when it is empty.

The interface and functionality need to change to improve visual appeal, ensure consistent branding, and prevent application crashes when the task list is empty.

Let's apply Lakers colors (purple and gold) in main-window.css and dialog-box.css.
Let's set a custom icon in the application window using stage.getIcons().add(new Image(...)).
Let's modify displayToDo to handle when the list is empty.
Let's add a Ui display to showcase latest display.

Using a Lakers-themed design aligns the UI with the brand identity.
Explicit error handling for empty tasks enhances robustness.
Customizing the window icon and tweaking layout details ensures a polished, user-friendly experience.

ho-thanh-und committed:
Merge pull request #26 from ho-thanh-und/branch-UI

UI improvements: add Lakers styling, set window icon, handle empty tasks

ho-thanh-und committed:
Event formatting: update event parsing and storage

The event tasks use an incorrect format that only supports a single time value, causing confusion and incomplete information when events span a time range. This limits users from specifying both a start and an end time.

Let's update Event.java to parse a period string in the format "<start> to <end>" and store separate start and end values, modify Storage.java to load and save events using the new period format, and update TaskList.java to parse user input accordingly.

This is done to provide a more natural, consistent, and flexible event format for both display and file storage.

ho-thanh-und committed:
Update README.md

ho-thanh-und committed:
Merge pull request #27 from ho-thanh-und/branch-A-UserGuide

Merge branch 'branch-A-UserGuide' to the Master branch

ho-thanh-und committed:
Update README.md
